Feature/registration login logout

maismin opened the pull request:
- Add API endpoint to register, login, and logout user
- Add authentication middleware

I moved the password encryption, matching, and jwt token retrieval to the User model since these methods revolve around the User. Would separating these methods into their own file have been better?

maismin committed:
Fix documentation

- Add periods to sentences in documentation

maismin committed:
Add bcryptjs and jsonwebtoken dependencies

maismin committed:
Add User model

maismin committed:
Fix env template to include jwt config variables

maismin committed:
Add an asyncHandler middleware

This reduces the need to include a try/catch for every API route that
uses the async/await syntax.

maismin committed:
Add authentication middle for protected routes

maismin committed:
Add API endpoints for to register, login, and logout user

aeciorc commented on server/app/models/User.js:
```diff
@@ -0,0 +1,63 @@
+  },
+  options,
+);
+
```
I think it would be a good idea to have [custom error handling](https://mongoosejs.com/docs/middleware.html#error-handling-middleware) here so that you can pass user-friendly messages to the frontend

E.g "The email entered is already taken" instead of "duplicated field value entered"

aeciorc commented on server/app/controllers/auth.js:
```diff
@@ -0,0 +1,123 @@
+ */
+// eslint-disable-next-line no-unused-vars
+const logout = asyncHandler(async (req, res, next) => {
+  res.cookie('token', 'none', {
```
there's also a `res.clearCookie(name)` you can use

aeciorc reviewed (approved):
Looks great, two suggestions below

maismin committed:
Fix express error middleware

- Add an extra condition for checking duplicate field values

maismin committed:
Add custom text for password validation

maismin committed:
Add custom error message for duplicate email

Esodey reviewed (approved):
All looks great

maismin committed:
Refactor logout endpoint

- Clear token in cookie immediately instead of refresh token with a
short expiration date and set to none.
